src/main/java/netzbegruenung/keycloak/authenticator/PhoneValidationRequiredAction.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright 2016 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * @author <a href="mailto:bill@burkecentral.com">Bill Burke</a>
+ * @author Niko Köbler, https://www.n-k.de, @dasniko
+ * @author Netzbegruenung e.V.
+ * @author verdigado eG
+ */
+
+package netzbegruenung.keycloak.authenticator;
+
+import netzbegruenung.keycloak.authenticator.gateway.SmsServiceFactory;
+
+import org.jboss.logging.Logger;
+import org.keycloak.authentication.CredentialRegistrator;
+import org.keycloak.authentication.RequiredActionContext;
+import org.keycloak.authentication.RequiredActionProvider;
+import org.keycloak.common.util.SecretGenerator;
+import org.keycloak.credential.CredentialModel;
+import org.keycloak.credential.CredentialProvider;
+import org.keycloak.models.AuthenticatorConfigModel;
+import org.keycloak.models.UserCredentialModel;
+import org.keycloak.models.UserModel;
+import org.keycloak.sessions.AuthenticationSessionModel;
+import org.keycloak.theme.Theme;
+import org.keycloak.util.JsonSerialization;
+
+import java.util.Locale;
+import java.util.Optional;
+import java.io.IOException;
+import javax.ws.rs.core.Response;
+
+public class PhoneValidationRequiredAction implements RequiredActionProvider, CredentialRegistrator {
+	private static final Logger logger = Logger.getLogger(PhoneValidationRequiredAction.class);
+	public static final String PROVIDER_ID = "phone_validation_config";
+
+	@Override
+	public void evaluateTriggers(RequiredActionContext context) {
+	}
+
+	@Override
+	public void requiredActionChallenge(RequiredActionContext context) {
+		context.getAuthenticationSession().addRequiredAction(PhoneNumberRequiredAction.PROVIDER_ID);
+		try {
+			UserModel user = context.getUser();
+			AuthenticatorConfigModel config = context.getRealm().getAuthenticatorConfigByAlias("sms-2fa");
```

Comment:
We need a more general approach here, as we cannot guarantee that this is always this ID.

Reply:
I was thinking that we could provide the config alias with auth notes. However, the processAction method of SmsAuthenticatorSetMobileNumberAction does not know the corresponding config as well. So we're back to square 1.

Reply:
We can use it for now for us. But we should probably change this. To merge this PR, we could state the caveat in the README that there needs to be one Execution with the name "sms-2fa" and working SMS API.

Reply:
Another dirty hack, but probably less nasty than other solutions, would be getting a config alias from the SQL database for the current realm.

Reply:
> Another dirty hack, but probably less nasty than other solutions, would be getting a config alias from the SQL database for the current realm. Not sure though if the database can be accessed in the scope of an auth provider.

Sounds good, I will try to fetch the config from DB.